### PR TITLE
Per-queue sequential cutoff support

### PR DESCRIPTION
**Pull request by robertbaldyga**
Requires https://github.com/Open-CAS/ocf/pull/448

Requires #721 due to order in which patches were merged in OCF

### CODE REVIEW

File: casadm/cas_main.c [hunk 1]
```diff
@@ -564,6 +564,13 @@ static struct cas_param cas_core_params[] = {
 		.name = "Sequential cutoff policy",
 		.value_names = seq_cutoff_policy_values,
 	},
+	[core_param_seq_cutoff_promotion_threshold] = {
+		.name = "Sequential cutoff promotion threshold [KiB]" ,
```
Comment (arutk) [inline casadm/cas_main.c:568]: I suggest "Sequential cutoff promotion size threshold [KiB]"

File: casadm/cas_main.c [hunk 2]
```diff
@@ -564,6 +564,13 @@ static struct cas_param cas_core_params[] = {
...
+		.transform_value = seq_cutoff_threshold_transform,
+	},
+	[core_param_seq_cutoff_promotion_count] = {
+		.name = "Sequential cutoff promotion request count" ,
```
Comment (arutk) [inline casadm/cas_main.c:572]: I suggest "Sequential cutoff promotion request count threshold"

File: casadm/cas_main.c [hunk 3]
```diff
@@ -632,6 +639,8 @@ static struct cas_param cas_cache_params[] = {
 #define SEQ_CUT_OFF_THRESHOLD_DESC "Sequential cutoff activation threshold [KiB]"
 #define SEQ_CUT_OFF_POLICY_DESC "Sequential cutoff policy. " \
 	"Available policies: {always|full|never}"
+#define SEQ_CUT_OFF_PROMO_THRESHOLD_DESC "Sequential cutoff stream promotion threshold [KiB]"
+#define SEQ_CUT_OFF_PROMO_COUNT_DESC "Sequential cutoff stream promotion request count"
```
Comment (arutk) [inline casadm/cas_main.c:643]: I suggest to change as per line 568 and line 572 comments

File: casadm/cas_main.c [hunk 4]
```diff
@@ -741,6 +752,18 @@ int set_param_seq_cutoff_handle_option(char *opt, const char **arg)
 			cas_printf(LOG_ERR, "Error: Invalid policy name.\n");
 			return FAILURE;
 		}
+	} else if (!strcmp(opt, "promotion-threshold")) {
+		if (validate_str_num(arg[0], "sequential cutoff promotion threshold", 0,
+				4194181) == FAILURE)
```
Comment (arutk) [inline casadm/cas_main.c:757]: did you mean 4194303 (2^(32 - 10) - 1) ?

File: casadm/cas_main.c [hunk 5]
```diff
@@ -741,6 +752,18 @@ int set_param_seq_cutoff_handle_option(char *opt, const char **arg)
...
+			return FAILURE;
+
+		SET_CORE_PARAM(core_param_seq_cutoff_promotion_threshold, atoi(arg[0]) * KiB);
+	} else if (!strcmp(opt, "promotion-count")) {
+		if (validate_str_num(arg[0], "sequential cutoff promotion request count", 0,
+				65535) == FAILURE)
```
Comment (arutk) [inline casadm/cas_main.c:763]: why not use UINT32_MAX?

File: casadm/cas_main.c (hunk 4 above)
Comment (arutk) [reply casadm/cas_main.c:757]: actually we could allow 2^32-1 if we didn't convert it to KiBs in the following lines

**Review by arutk — CHANGES_REQUESTED** [review]

**Comment by arutk** [comment]
there is a typo in PR title ("pre"- > "per")

File: casadm/cas_main.c (hunk 1 above)
Comment (robertbaldyga) [reply casadm/cas_main.c]: N/A

File: casadm/cas_main.c (hunk 4 above)
Comment (robertbaldyga) [reply casadm/cas_main.c]: N/A
Btw this number (which is also cutoff threshold limit) was historically introduced as an easter egg. It's the largest prime number smaller than 2^32.

File: casadm/cas_main.c (hunk 5 above)
Comment (robertbaldyga) [reply casadm/cas_main.c:753]: Per-stream request count is 16 bits long. We could technically make it bigger, but 2^16 is more than enough for promotion count, which will typically be set to value in the order of tens.

File: casadm/cas_main.c (hunk 3 above)
Comment (robertbaldyga) [reply casadm/cas_main.c]: Done.

File: casadm/cas_main.c (hunk 2 above)
Comment (robertbaldyga) [reply casadm/cas_main.c]: Done.

File: modules/cas_cache/layer_cache_management.c [hunk 6]
```diff
@@ -2184,6 +2260,54 @@ int cache_mngt_get_seq_cutoff_policy(ocf_core_t core,
...
+	result = ocf_mngt_core_get_seq_cutoff_promotion_threshold(core, thresh);
+
+	ocf_mngt_cache_read_unlock(cache);
+	return result;
+}
+
+/**
+ * @brief Get sequential cutoff promotion request count value
+ * @param[in] core OCF core
+ * @param[out] thresh sequential cutoff promotion request count value
```
Comment (rafalste) [inline modules/cas_cache/layer_cache_management.c:2228]: Did you mean 'count' instead of 'thresh'?

Comment (robertbaldyga) [reply modules/cas_cache/layer_cache_management.c]: Yep. Fixed.

**Comment by robertbaldyga** [comment]
retest this please

**Review by arutk — APPROVED** [review]

**Review by mmichal10 — APPROVED** [review]

**Pull request merged**